### PR TITLE
fix(npm): align Node esm code importing cjs with Node

### PR DESCRIPTION
**Pull request by dsherret**
This more aligns Node ESM importing CJS to have the same behaviour as Node, but keeps the current `__esModule` check that Deno does when importing CJS from npm. Essentially, this allows the Node ESM code to be comptabile with Node and the Deno code can be more comptable with what's done in the JS/TS ecosystem. That said, I suspect we may run into issues doing this though and may want to eventually align Deno's interop with Node's behaviour...

### CODE REVIEW

File: cli/module_loader.rs [hunk 1]
```diff
@@ -134,30 +137,52 @@ impl CliModuleLoader {
     maybe_referrer: Option<ModuleSpecifier>,
   ) -> Result<ModuleSource, AnyError> {
     let code_source = if self.ps.npm_resolver.in_npm_package(specifier) {
-      let file_path = specifier.to_file_path().unwrap();
+      let is_cjs = self.ps.cjs_resolutions.lock().contains(specifier);
```
Comment (littledivy) [inline cli/module_loader.rs:140]: Question: Why is `cjs_resolutions` a Mutex? Is it shared across threads? Assuming not, RefCell should be much cheaper.

Comment (dsherret) [reply cli/module_loader.rs:140]: Yeah, the data in ProcState is shared across threads.

File: cli/module_loader.rs [hunk 2]
```diff
@@ -134,30 +137,52 @@ impl CliModuleLoader {
...
+        {
+          let mut specifier = specifier.clone();
+          specifier.set_path(new_path);
+          (Some(CjsToEsmTranslateKind::Node), Cow::Owned(specifier))
+        } else if let Some(new_path) =
+          specifier.path().strip_suffix(node::CJS_TO_ESM_DENO_SUFFIX)
+        {
+          let mut specifier = specifier.clone();
+          specifier.set_path(new_path);
+          (Some(CjsToEsmTranslateKind::Deno), Cow::Owned(specifier))
```
Comment (littledivy) [inline cli/module_loader.rs:153]:
```suggestion
        let mut specifier = specifier.clone();
        let path = specifier.path();
        if let Some(new_path) =
          path.strip_suffix(node::CJS_TO_ESM_NODE_SUFFIX)
        {
          specifier.set_path(new_path);
          (Some(CjsToEsmTranslateKind::Node), Cow::Owned(specifier))
        } else if let Some(new_path) =
         path.strip_suffix(node::CJS_TO_ESM_DENO_SUFFIX)
        {
          specifier.set_path(new_path);
          (Some(CjsToEsmTranslateKind::Deno), Cow::Owned(specifier))
```

Comment (littledivy) [reply cli/module_loader.rs:153]: if that compiles :). LGTM otherwise

**Review by littledivy — APPROVED** [review]

File: cli/module_loader.rs (hunk 2 above)
Comment (dsherret) [reply cli/module_loader.rs]: I wish the url crate had "with" methods, so you could do `url.with_path(path)` to create a new url with that path.

**Pull request merged**